### ISSUE TITLE
When to compute?

### ISSUE THREAD

**Issue by saulshanabrook** [Quansight-Labs/uarray#12]
I am working on replicating the paper's example in uarray. I made some good progress here (https://github.com/Quansight-Labs/uarray/blob/6e03be28383fdc08bef02cd0f6662614e69f3d05/uarray/Intro.ipynb), but realize that I need to change how I am processing concrete arrays. Right now, for things like equivalence, I compute all possible indexes of the array and see if every element is equal. That relies on an `ExplodeVector` operation which takes a 1D array and returns a new 1D array with every element indexed so each can be replaced. I am realizing now that this approach breaks down and becomes confusing, because you don't know when exactly to call `ExplodeVector`.

The underlying problem is when to actually compute intermediate results. When we know an arrays contents or shape at compile time, we should use that eagerly and compute operations on that as we replace. Why? Because we need this for shape calculations. What if you don't want to use some data at compile time? Like if you have big numpy arrays, you don't want it computing things willy nilly on them while it is pattern matching. You want it to wait and return some sort of expression that you can use to compute the result. I think this is the point @hameerabbasi was making last week when we talked about this (correct me if I am am remembering incorrectly).

The form in the paper goes from this originally:

```
((<1 0> ψ A^3) +·× (<2> ψ ((<1 0> ψ A^3) ·× (<0 1> ψ B^3))))
```

To this replaced:

```
((<1 0> ψp A^3) +·× (((<1 0> ‡v (<1> ↑ <2>)) ψf A^3) ·× ((<0 1> ‡v (<1> ↓ <2>)) ψp B^3)))
```

(`ψp` is just the index operator when we know it's a partial index vs a full index. `‡v` is vector concat)

Ok so we see this expression `(<1> ↑ <2>)` in there which should be simplified. It should just return `<2>`, because this is saying "Take the first 1 item in the vector". So we could go and add a manual rule for Take when the arguments are two concrete vectors, which would work. But this is superfluous. We have already defined `Take` in terms of shape and index, and the resulting concrete transformation can be inferred from this. Right now, the I would call `ExplodeVector(vector(1), expr)` to transform this, since this would be replaced by indexing on that original expression when would then be replaced by the `Index` rule for `Take`. 

But when do I call `ExplodeVector`? It seems arbitrary. I don't want to have to sprinkle this all over. I don't even have a coherent reason for where I should put it. So we should do things another way.

OK so let's see if we can come up with a plan to remove uses of `ExplodeVector`.

One thing that could help is to use abstract arrays. I used these when defining this expression, as the `A` and `B` terms, with `A = AbstractWithDimension(Scalar(3), variable_name="A")`. One place where `ExplodeVector` is used is in `Equiv`. Instead of seeing if every element of the two arrays is the same, we should see if there shapes have the same form and if indexing them has the same form. We can use the `__eq__` method on the actual operations to compare equivalence of form. It recursively verifies that the trees are equal. But... You could have two different expressions that have the same normal form that are not equal. A simple example is that `Take(vector(1), vector(10))` is equivalent to `vector(10)`. So I think in this case, for `Equiv` we can fully index each expression by a vector of abstract scalars. Then this will cause the pattern matching for Index to take place and we will get the normal form.

This is how Lenore simplifies the expressions in the paper. She indexes them with a vector of `<i>`, so that the rules are applied. 

OK but what about something like `Take(vector(0), Scalar(10))`? Here the dimensionality is 0 so we don't have to index it. Yet, indexing it would cause Take's index replacement to be used so would be necessary to cause it to be simplified. So we would have to do `Index(Vector(), Take(vector(0), Scalar(10))` which is a bit weird because you also want to remove indexing by an empty vector, i.e. replace it with just it's result. In which case the indexing of take wouldn't be applied. Now yes, this could be fixed by adding multiple matching levels, so the take indexing is guaranteed to be replaced before the replacement rule of indexing by an empty vector, but this seems very ad hoc.

So this makes me think that any operation that results in a scalar should just be replaced by that definition. For example, the cross product of two scalars should be replaced by the resulting scalar (even if that scalar is not a concrete values but some operation). It seems silly to wait for an empty index to trigger this replacement.

So this brings us to needing shape information when defining the indexing. For example, we wanna say something like "when the result of the cross product is a scalar, replace it with that result. Otherwise keep it"

It makes me wonder about changing our representation entirely into three different abstract array operations:

```python
AbstractWithDimensions(n_dim)
AbstractWithShape(n_dim, shape)
AbstractWithIndex(n_dim, shape, index_fn)
```

We would also keep some sort of concrete vector (`Vector`) and concrete scalar (`Scalar`). 

This would make the replacements a bit clearer, so that you are replacing OuterProduct with its normal form, instead of defining replacements on taking the shape or index of it. 

So the million dollar question is how to represent the `index_fn` symbolically. It could be represented as another array? With abstract array being the indexes? OK so how would we define the replacement for  `Index(x, AbstractWithIndex(n_dim, shape, index_fn))`?

Not sure, but this form would allow us to do something like `AbstractWithIndex(Scalar(0), x, index_fn) -> index_fn`.

We could define it as the resulting array with the unknown indices being `AbstractWithDimensions(0, variable_name=f"index_{i}")` (scalars). That's all well and good but how do you replace those abstract dimensions with real ones once you index? You could add some bound scoping. Like a `WithVariables(Scalar({"index_1": xxx}), expr)` operation that is replaced by forwarding on the WithVariables to each of the expressions. Something like this:

```python
register(WithVariables(scalar, x), lambda scalar, x: matchpy.functions.substitute(scalar.value, x))
```

I found that matchpy actually [has a substitute function already defined](https://matchpy.readthedocs.io/en/latest/api/matchpy.functions.html#matchpy.functions.substitute) that seems like it does exactly what we want! OK this makes me feel like maybe this is an OK track to go down then.  (maybe this should be called `Substitute` then instead of `WithVariables`). 

So how about we assign the index vector to a variable, like this:

```python
register(Index(x, AbstractWithIndex(x1, x2, x3)), lambda x, x1, x2, x3: substitute({"index": x}, x3))
```

However, then all indexes will be labeled the same. OK so let's add a parameter to `AbstractWithIndex` which is the name of the index variable, so it works like `AbstractWithIndex(dim, shape, index_variable_name, indexed_value)` where `index_variable_name` is a scalar of a string.

This sounds OK to try to implement. I wanna rename them though to this:

```
AbstractWithDimensions -> HasDim(dim)
AbstractWithShape -> HasShape(shape)
AbstractWithIndex -> HasIndex(shape, index_variable_name, contents)
```

Then maybe:

```
Scalar -> HasValue
Vector -> HasItems
```

**Comment by pearu** [comment]
The rename targets sound like predicates. Are they?

**Comment by saulshanabrook** [comment]
Good point. No they are not.

**Comment by saulshanabrook** [comment]
I could make them Symbol's instead of Operation's maybe. Then they would be leaves in the tree instead of branches and their arguments wouldn't be replaced on. Hm... Maybe `WithDim`, etc makes more sense as well.

**Comment by saulshanabrook** [comment]
If they are symbols they are opaque to MatchPy, like https://matchpy.readthedocs.io/en/latest/example.html#symbols

**Comment by saulshanabrook** [comment]
> Do we need the shape or just the number of dimensions at the time of compilation?

Sometimes one, sometimes the other.

**Comment by hameerabbasi** [comment]
Then let's make it shape, with the possibility of having a generic integer placeholder for a dimension. We might learn from numpy/numpy-stubs#5 here.

**Comment by pearu** [comment]
Btw, perhaps offtopic for this issue, when thinking about also ragged arrays, the shape and dimension that represented as a vector of indices and an integer might be too restrictive. I think the shape is the most fundamental concept in MoA (for instance, the dimension can be derived from it when well-defined (for ragged arrays it might not be, e.g. for array `[[a, b], c]`)). The shape basically defines what are valid indices for arrays for which the items exist. This holds for all arrays: multidimensional, sparse, ragged.
The dimension could be defined as the length of the longest valid index vector.
I guess my message here is that, if possible, allow other representations of shape as well, not only the representation where the shape is a vector of indices.

**Comment by saulshanabrook** [comment]
Thanks @hameerabbasi for the pointer to that discussion on numpy typing, which is similar to this one. We aren't trying to achieve any static analysis, so we have a bit more flexibility than with type annotations. But merging this effort with python typing would be nice at some point!  

---

> I guess my message here is that, if possible, allow other representations of shape as well, not only the representation where the shape is a vector of integers.

Good point! I would think that the mathematical equivalences of indexing  and shape calculus presented in MoA thesis would need to be updated to be compatible with other shape representation. So when we do start supporting other types of shapes, we just need to make sure the existing definitions are not applied where they are not valid. 

--- 

After a day mulling over my proposed solution here, I am not sure if it will help the situation. But I did think of a few things to clean up fundamentally that will hopefully help the situation. I opened https://github.com/Quansight-Labs/uarray/issues/13 to articulate these.
